### PR TITLE
Fix issues found by static checkers

### PR DESCRIPTION
**Pull request by ejona86**
Unused variables, improper GuardedBy, places where static could be
added.

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/ServerImpl.java [hunk 1]
```diff
@@ -81,7 +81,7 @@
 
   /** Executor for application processing. */
   private Executor executor;
-  @GuardedBy("lock") private boolean usingSharedExecutor;
+  private boolean usingSharedExecutor;
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: Just lock usingSharedExecutor down in shutdown()

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 2]
```diff
@@ -665,7 +665,7 @@ public String getLogId() {
     return GrpcUtil.getLogId(this);
   }
 
-  private class NameResolverListenerImpl implements NameResolver.Listener {
+  private static class NameResolverListenerImpl implements NameResolver.Listener {
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: whats the issue?

File: core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java [hunk 3]
```diff
@@ -205,15 +204,14 @@ public void enterIdleModeAfterForceExit() throws Exception {
     newTransports.drainTo(transports);
     assertEquals(addressGroupList.size(), transports.size());
 
-    InterimTransport<ClientTransport> interimTransport = channel.tm.createInterimTransport();
+    channel.tm.createInterimTransport();
```
Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java]: why does this return anything?

File: core/src/main/java/io/grpc/internal/ServerImpl.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: Well, it is really the same as `executor`, which isn't `@GuardedBy` either. I'd rather they stay the same.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 2 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: No need for non-static.

File: core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java (hunk 3 above)
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java]: I'm not sure I understand the question. The method returns a new interim transport. In this case, the test just needed to create it. I believe it is registered within `ManagedChannelImpl`, so it's not like we're just doing a `new` here to ignore the result.

File: core/src/main/java/io/grpc/internal/ServerImpl.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: Ok, add `@GuardedBy`  

Also, it looks like SharedResourceHolder.release is not nulling out the executor below.

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: Adding `@GuardedBy` to `executor` seems an overkill and not something we'd want to do. `executor` is used on a fast-path that currently is not using `lock`. Simply put, neither variable is or should be guarded by the lock.

It actually looks like the executor can be used after releasing (I just created #2034) which means we can't set it to null until that is fixed first (and even then I'm not sure if it is worth-while).

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: Then remove it from the synchronized block?

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: It must be set before `transportServer.start()` (which means it couldn't be removed from the synchronized block even if it didn't need it) and it needs some coordination with shutdown(). It needs synchronization, but isn't the GuardedBy semantics.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: I don't feel comfortable removing documentation that describes how variables should be properly synchronized.  This needs a comment (compiled as annotation or otherwise) describing how to safely access executor, usingSharedExecutor, and the other variables in this class.  

I don't think moving usingSharedExecutor and executor to inside the synchronized block is overkill.  What bad thing is going to happen?

File: core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java (hunk 3 above)
Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java]: Looking at InterimTransport, wouldn't be better to keep the ref and close it at the end of this method?

File: core/src/main/java/io/grpc/internal/ServerImpl.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: > I don't feel comfortable removing documentation that describes how variables should be properly synchronized.

You _just_ added it in #2011. But it was wrong. I have little issue removing wrong docs, especially as they break compilation internally.

> I don't think moving usingSharedExecutor and executor to inside the synchronized block is overkill. What bad thing is going to happen?

It is completely unnecessary. And it would be on the fast path.

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: I'd be more prone to reverting #2011, because it breaks the build.

File: core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java (hunk 3 above)
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java]: I don't see cleanup handling for any of these tests. I think they are all depending on normal GC.

**Comment by ejona86** [comment]
Please see the new commit for cancelIdleTimer. This is now the only blocker for syncing.

File: core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java (hunk 3 above)
Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java]: I view tests as being examples/documentation how code should be used.  Since the real code would do it, it should be here.

Comment (ejona86) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java]: These tests aren't sharing that view with you. They don't shutdown anything. It's not this one line then that would need to be changed, but the entire file.

File: core/src/main/java/io/grpc/internal/ServerImpl.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: All the more reason that the synchronization of the variable needs to be clearly specified.  This class is so poorly documented that it requires considerable effort to understand.  

Either add it to the synchronized block, or comment that it is explicitly not synchronized.  As of right now, it looks like it was just forgotten.

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: So you're saying you forgot executor as well in #2011? How is this now my job? I'm going to revert #2011.

File: core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java (hunk 3 above)
Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java]: Ok, then fix them?  PRs should make code better.  Either leave this line alone, or fix the code to be closer to ideal.

Comment (ejona86) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplIdlenessTest.java]: That is your own personal preference and it does not need to apply to these tests. I can see the value, but I don't think it is a universal rule, since many tests have to do lots of stuff I would never want a real user to do. Preventing a single one-line change that is consistent with the class because the class was not written as you would like does not seem acceptable.

File: core/src/main/java/io/grpc/internal/ServerImpl.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: I figured out a short comment.

**Comment by carl-mastrangelo** [comment]
If you look at the reason that annotation was added in the first place, it was because EP found a unsynchronized error with commit c6fd94ca8537035a4dc9a6c51662724c1a172266, introduced by you.

I don't mean to assign blame, but it is clear to me that both of us have misunderstood how the variables in the class should be synchronized.  This is the second time, which Is why I am asking for this.  It is no longer hypothetical nor a fluke.

**Comment by ejona86** [comment]
@carl-mastrangelo, did you see the comment I added?

> I don't mean to assign blame, but it is clear to me that both of us have misunderstood how the variables in the class should be synchronized.

Not really. Look at it. It was a typo. I don't disagree that we should have better documentation describing threading, but blocking this PR on it did not seem appropriate. As I showed `executor` was already missing documentation. The PR without a comment was strictly better than the previous code since now it could compile with ErrorProne while being no more nor less correct.

**Comment by pgrosu** [comment]
Hi Carl and Eric,

It took me more than a year just to understand many of the intricacies of [gRPC from the code](https://github.com/grpc/grpc), since the documentation sometimes was too sparse and the code-changes were sometimes frequent without a clear, step-wise design specification plan.  This is no fault of the team, which is fantastic and helpful, but very overwhelmed.  When I create an API, I usually go a little more extreme where I begin with all the tests - including all edge cases - before starting to code anything, in order to ensure the requirements are satisfied through a test-driven development approach.  It is only after that, that I write my function contracts which becomes a mechanical process, because the tests tell me what to provide and expect.  This got burned in my mind after many painful code reviews.

I definitely agree with Carl on this one.  Code that is not properly documented leads to even more confusion.  You have created a great implementation here and there is no rush to just push something if it is not properly ready, which in my eyes as a user means the documentation of the API - and as an end-user would think is the most pertinent.  It has always saved me many headaches down the line, as sometimes I usually include more documentation than code for tricky parts of the code.

We are all a team here to satisfy the users of this great API, so I think it is a shared job.

Hope you share my perspective, and it is not too much to ask.

Thank you,
Paul

**Comment by ejona86** [comment]
@pgrosu, I may agree with you and @carl-mastrangelo may agree with you, but this test is already written and not by either of us. As a team we don't require tests to be written that way though, and I know that some of mine it is not feasible to write them so as to provide documentation. I also think that you _shouldn't need to read tests_ in order to mostly understand a class. It just happens that when documentation sucks, tests turn out to be helpful.

But rewriting multiple tests it in order to remove an used variable variable is outrageous to me, especially when we don't agree as a team that tests are to be written as if they were documentation.

**Comment by ejona86** [comment]
I've split out some of these changes into #2040, to allow the test changes discussion to continue.

**Comment by pgrosu** [comment]
Hi Eric,

I understand your frustration, and I had to justify myself at times regarding how I would change inherited code.  I even had to write placeholder functions to enable existing functionality to continue, since users wanted to use both older and newer features that conflicted with the previous implementation.  You are actually luckier than most since the version is pre-1.0, so there is room to change the path even if it might be harder, and less traveled by.

Think of it from the point of view of an end-user.  When will an API bring one joy to use?  When there is purpose to each element of the code and becomes intuitive to use, which is enabled through proper documentation and validated by tests.  It is a euphoric feeling when things just work and make sense, and also has full support of the development team.

A piece of code usually stems from a behavior-driven development specification plan, that matches the usage requirements, where the flow of information is detailed.  If that does not exist, then a test-driven development process follows to provide minimum satisfiability requirements of functionality.  Each test would have a documented purpose, so as to link them accordingly and allow for proper class/function contract definitions.  If those exist but the code requires changes and the tests stay the same - and/or might not exist - then there is a dichotomy between interpreting how the code operates and what the implementation goals are from the development team.  This can frustrate end-users which are sometimes not easily adopted.

The easiest approach for simple code-changes would be to ensure everyone is on the same page regarding how all the parts work together, and then correct them appropriately with all key folks which should be part of the process.  Sometimes to bring everyone on the same page through an open discussion, this requires to take a step back and write the current understanding and proposed implementation down, and present it via an issue to properly hash out ideas.  We are in the middle of that discussion here, but it might be helpful to write down how all the impacted parts work together and how the proposed change will allow for an improved implementation affecting those parts.

After reaching agreement on the plan forward and initiating the updates, all the code-changes should satisfy the requirement specifications, and the tests should be updated where necessary to validate those specs.  Any additional documentation that allows for users of different levels to have a welcome experience in using the API will continue the growth a fruitful community.

Hope you see where I am coming from, and that the above would be points of agreement even if they might make things a little tougher in the short term to work through.  I welcome a discussion on any of them.

Thanks,
Paul
